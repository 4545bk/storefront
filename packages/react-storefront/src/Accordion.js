/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { useState, useContext } from 'react'
import AppContext from './AppContext'
import AmpAccordion from './amp/AmpAccordion'

const AccordionContext = React.createContext()

export const withAccordionContext = Component => {
  return props => (
    <AccordionContext.Consumer>
      {ctx => <Component {...props} {...ctx} />}
    </AccordionContext.Consumer>
  )
}

/**
 * Accordion which only allows one child ExpandableSection to be open at a time
 *
 * ```js
 *  <Accordion>
 *    <ExpandableSection title="First">
 *      <div>The first section</div>
 *    </ExpandableSection>
 *    <ExpandableSection title="Second">
 *      <div>The second section</div>
 *    </ExpandableSection>
 *    <ExpandableSection title="Third">
 *      <div>The third section</div>
 *    </ExpandableSection>
 *  </Accordion>
 * ```
 */
export default function Accordion({ children, ...otherProps }) {
  const [expandedSectionId, setExpandedSectionId] = useState(null)
  const {
    app: { amp }
  } = useContext(AppContext)
  return (
    <AccordionContext.Provider
      value={{
        expandedSectionId,
        setExpandedSectionId
      }}
    >
      {amp ? <AmpAccordion {...otherProps}>{children}</AmpAccordion> : children}
    </AccordionContext.Provider>
  )
}
