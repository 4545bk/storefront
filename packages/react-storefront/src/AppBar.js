/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { throttle } from 'lodash'
import Toolbar from '@material-ui/core/Toolbar'
import SvgIcon from '@material-ui/core/SvgIcon'
import { observer, inject } from 'mobx-react'
import Hidden from '@material-ui/core/Hidden'
import withStyles from '@material-ui/core/styles/withStyles'
import { Vbox } from './Box'
import ToolbarButton from './ToolbarButton'
import MenuIcon from './assets/menu.svg'

export const styles = (theme) => ({
  root: {
    boxSizing: 'content-box',
    position: 'relative',
    zIndex: theme.zIndex.modal + 10
  },

  withAmp: {
    zIndex: theme.zIndex.amp.modal + 1
  },

  icon: {
    color: theme.palette.action.active
  },

  toolBar: {
    height: theme.headerHeight || '64px',
    padding: `0 7px`,
    maxWidth: theme.maxWidth,
    flex: 1
  },

  wrap: {
    borderBottom: `1px solid ${theme.palette.divider}`,
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    backgroundColor: theme.palette.background.paper,
    zIndex: theme.zIndex.modal + 10,
    display: 'flex',
    justifyContent: 'center',
  },

  stuck: {
    transform: 'translateY(0)',
  },

  unstuck: {
    transform: 'translateY(-100%)',
  },

  animate: {
    transition: 'transform .15s ease-in',
  },

  hidden: {
    position: 'fixed',
    zIndex: theme.zIndex.modal + 10,
    boxShadow: theme.shadows[2],
    top: 0,
    left: 0,
    right: 0
  },

  menuOpen: {
    boxShadow: 'none'
  },

  link: {
    textDecoration: 'none'
  }
});

/**
 * A header that auto hides when the user scrolls down and auto shows when the user 
 * scrolls up. A hamburger button that controls the menu is automatically displayed on
 * the left side of the header.  Children are placed directly to the right of the menu
 * button.
 */
@inject(({ app }) => ({ menu: app.menu, amp: app.amp }))
@withStyles(styles, { name: 'RSFAppBar'})
@observer
export default class Header extends Component {

  state = {
    stuck: false,
    hidden: false,
    animate: false
  }

  static propTypes = {
    /**
     * CSS classes
     */
    classes: PropTypes.object,

    /**
     * Set to true to show the hamburger button only on mobile devices. Defaults to false.
     */
    responsive: PropTypes.bool,

    /**
     * A component for the menu icon.
     */
    MenuIcon: PropTypes.func
  }

  static defaultProps = {
    responsive: false,
    MenuIcon
  }

  render() {
    const { MenuIcon, classes, children, fixed, menu, responsive, width, amp, showLabels } = this.props

    const menuIcon = <SvgIcon className={classes.icon}><MenuIcon/></SvgIcon>

    return (
      <div className={classnames({ [classes.root]: true, [classes.withAmp]: amp })}> 
        <div className={classnames({
          [classes.wrap]: true,
          [classes.fixed]: fixed,
          [classes.hidden]: this.state.hidden || menu.open,
          [classes.stuck]: (this.state.hidden && this.state.stuck) || menu.open ,
          [classes.unstuck]: this.state.hidden && !this.state.stuck,
          [classes.animate]: this.state.animate,
          [classes.menuOpen]: menu.open
        })}>
          <Toolbar disableGutters classes={{ root: classes.toolBar }}>
            <Hidden mdUp={responsive} implementation="css">
              <a on="tap:moov_menu.toggle" className={classes.link}>
                <ToolbarButton 
                  aria-label="Menu" 
                  color="inherit" 
                  onClick={menu.toggle}
                  icon={menuIcon}
                />
              </a>
            </Hidden>
            { children }
          </Toolbar>
        </div>
      </div>
    )
  }

  componentDidMount() {
    window.addEventListener('scroll', this.onScroll, { passive: true })
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.onScroll, { passive: true })
  }

  componentDidUpdate(prevProps, prevState) {
    if (!prevState.hidden && this.state.hidden) {
      setTimeout(() => this.setState({ animate: true }), 100)
    }
  }

  onScroll = () => {
    const height = 64, { scrollY } = window, { lastScrollY } = this
    const { menu } = this.props

    if (this.state.hidden) {
      if (scrollY <= 0) {
        this.setState({ hidden: false, stuck: false, animate: false })
      }
    } else {
      if (scrollY > height && !menu.open) {
        this.setState({ hidden: true })
      }
    }

    const unstickBufferZone = 15

    if (this.throttledScrollY > scrollY + unstickBufferZone && !this.state.stuck) {
      this.unstickAt = scrollY + unstickBufferZone
      this.setState({ stuck: true })
    } else if (scrollY > this.unstickAt && this.state.stuck && !menu.open) {
      this.setState({ stuck: false })
      delete this.unstickAt
    }

    if (this.lastScrollY > scrollY && this.state.stuck) {
      this.unstickAt = scrollY + unstickBufferZone
    }

    this.lastScrollY = scrollY
    this.sampleScrollSpeed()
  }

  sampleScrollSpeed = throttle(() => this.throttledScrollY = window.scrollY, 100)

}
