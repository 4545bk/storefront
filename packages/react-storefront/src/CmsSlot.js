/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component }  from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import withStyles from '@material-ui/core/styles/withStyles'

export const styles = theme => ({
  inline: {
    display: 'inline'
  },
  block: {
    display: 'block'
  }
})

/**
 * A container for HTML blob content from a CMS.  Content is dangerously inserted into the DOM.
 * Pass the html as a string as the child of this component.
 */
@withStyles(styles)
export default class CmsSlot extends Component {
  static proptypes = {
    /**
     * Use inline prop to use display:inline style
     */
    inline: PropTypes.boolean
  }
  render() {
    const { children, className, classes, inline } = this.props
    return children ? <span className={classnames(className, { [classes.inline]: inline, [classes.block]: !inline })} dangerouslySetInnerHTML={{ __html: children }} /> : null
  }
}