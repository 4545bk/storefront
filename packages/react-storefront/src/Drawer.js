/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import ResizeObserver from 'resize-observer-polyfill'
import MUIDrawer from '@material-ui/core/Drawer'
import { withStyles } from '@material-ui/core/styles';
import Close from '@material-ui/icons/Close'
import Button from '@material-ui/core/Button'
import PropTypes from 'prop-types'
import Typography from '@material-ui/core/Typography'
import { inject } from 'mobx-react'
import classnames from 'classnames'

/**
 * A slide-in drawer with fab close button.
 */
export const styles = theme => ({
  closeButton: {
    position: 'absolute',
    right: '10px',
    top: '-28px',
    zIndex: 1
  },

  paper: {
    overflowY: 'visible'
  },

  title: {
    textAlign: 'center',
    padding: '20px',
    fontSize: '18px',
    borderBottom: `1px solid ${theme.palette.divider}`
  },

  ampClosed: {
    display: 'none'
  }
});

@withStyles(styles, { name: 'RSFDrawer' })
@inject(({ app, ampStateId }) => ({ amp: app && app.amp, ampStateId }))
export default class Drawer extends Component {

  static propTypes = {
    /**
     * Set to false to hide the close button. Defaults to true
     */
    showCloseButton: PropTypes.bool,

    /**
     * Called when the user closes the drawer
     */
    onRequestClose: PropTypes.func.isRequired,

    /**
     * The title to display at the top of the drawer
     */
    title: PropTypes.oneOfType([ PropTypes.string, PropTypes.element ]),

    /**
     * Set to true to automatically add padding to the body when the drawer
     * is open so that the user is able to scroll and see all of the body content.
     * Defaults to false.
     */
    autoAdjustBodyPadding: PropTypes.bool,

    /**
     * Props to apply to the closeButton
     */
    closeButtonProps: PropTypes.object,

    /**
     * The name of a property the amp-state to bind to the closed state of the drawer.
     */
    ampBindClosed: PropTypes.string
  }

  static defaultProps = {
    showCloseButton: true,
    autoAdjustBodyPadding: false,
    closeButtonProps: {}
  }

  constructor() {
    super();
    this.drawer = React.createRef()
  }

  render() {
    const { amp, ampStateId, ampBindClosed, variant, closeButtonProps, showCloseButton, open, onRequestClose, title, children, classes, autoAdjustBodyPadding, ...rest } = this.props

    return (
      <MUIDrawer 
        anchor="bottom" 
        classes={{
          paper: classnames({
            [classes.paper]: true,
            [classes.ampClosed]: amp && !open
          })
        }}
        amp-bind={ampBindClosed ? `class=>${ampStateId}.${ampBindClosed} ? '${classes.ampClosed}' : null` : null} 
        open={(amp && variant === "temporary") || open}
        variant={variant}
        {...rest}
      >
        <div ref={this.drawer}>
          { title && (
            <div className={classes.title}>
              <Typography variant="h6" component="div">{title}</Typography>
            </div>
          )}

          { showCloseButton && (
            <Button 
              variant="fab" 
              color="primary" 
              className={classes.closeButton} 
              onClick={this.closeDrawer} 
              style={{ display: open ? '' : 'none' }}
              on={ampBindClosed ? `tap:AMP.setState({ ${ampStateId}: { ${ampBindClosed}: true }})` : null}
              {...closeButtonProps}
            >
              <Close/>
            </Button>
          )}

          { children }
        </div>
      </MUIDrawer>
    )
  }

  // add body padding equal to drawer height
  componentDidMount() {
    this.setPadding()
    const el = this.drawer.current

    if (this.props.autoAdjustBodyPadding && el) {
      this.drawerResize = new ResizeObserver(() => {
        document.body.style.paddingBottom = el && el.clientHeight + 'px'
      })

      this.drawerResize.observe(el)
    }
  }

  componentWillUnmount() {
    const el = this.drawer.current

    if (this.drawerResize && el) {
      this.drawerResize.unobserve(el)
    }
  }

  // if value of open property changes, recalcuate drawer height/body padding
  componentWillReceiveProps(nextProps) {
    if (nextProps.open && !this.props.open) {
        this.setPadding()
    } else if (this.props.open && !nextProps.open) {
        this.closeDrawer()
    }
  }

  setPadding() {
    if (this.props.autoAdjustBodyPadding) {
      requestAnimationFrame(() => {
        const el = this.drawer.current
        document.body.style.paddingBottom = el && el.clientHeight + 'px'
      })
    }
  }

  closeDrawer = () => {
    const { onRequestClose } = this.props
    onRequestClose()
    document.body.style.paddingBottom = 0
  }

}
