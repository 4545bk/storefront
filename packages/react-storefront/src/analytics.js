/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */

let _targets = []

/**
 * Configures the PWA to broadcast events to all specified targets.  Once configured, You can all any method on
 * the the default import from this module it will attempt to call a method of the same name on each target.  
 * 
 * Example:
 * 
 *  // src/analytics.js
 * 
 *  import { configureAnalytics } from 'react-storefront/analytics'
 *  import GoogleTagManagerTarget from 'moov-pwa-analytics/GoogleTagManagerTarget'
 *  import MerkleSearchTarget from 'moov-pwa-analytics/MerkleSearchTarget'
 * 
 *  configureAnalytics(
 *    new GoogleTagManagerTarget({ apiKey: 'my_gtm_api_key' }),
 *    new MerkleSearchTarget({ apiKey: 'my_merkle_api_key' })
 *  )
 * 
 * // then, to broadcast an analytics event to all targets:
 * 
 * import analytics from 'react-storefront/analytics'
 * 
 * analytics.someEvent() // this will call the 'someEvent' method on all configured targets
 * 
 * @param {AnalyticsTarget[]} targets An array of targets to notify when analytics events occur
 * @return {AnalyticsProvider}
 */
export function configureAnalytics(...targets) {
  _targets = targets
}

export function getTargets() {
  return _targets
}

/**
 * This module's default export is a single entry point for broadcasting analytics events to all targets configured via `configureAnalytics`.
 * Each target represents a tag manager or other analytics service to which the application should send data.
 * 
 * Example:
 * 
 *  // src/analytics.js
 * 
 *  import { configureAnalytics } from 'react-storefront/analytics'
 *  import GoogleTagManagerTarget from 'moov-pwa-analytics/GoogleTagManagerTarget'
 *  import MerkleSearchTarget from 'moov-pwa-analytics/MerkleSearchTarget'
 * 
 *  configureAnalytics(
 *    new GoogleTagManagerTarget({ apiKey: 'my_gtm_api_key' }),
 *    new MerkleSearchTarget({ apiKey: 'my_merkle_api_key' })
 *  )
 * 
 */
export default new Proxy(
  { 
    setHistory: () => {} // polyfill for IE11 requires defined fields in Target argument
  }, 
  {
    get: function (o, method) {
      return function(...args) {
        for (let target of _targets) {
          const fn = target[method]

          if (fn) {
            fn.apply(target, args)
          } else {
            if (typeof method === 'string') {
              console.warn(`${target.constructor.name} does not support ${method}`)
            }
          }
        }
      }
    }
  }
)
