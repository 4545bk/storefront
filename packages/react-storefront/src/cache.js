/**
 * @license
 * Copyright Â© 2017-2019 Moov Corporation.  All rights reserved.
 */
import { clearCache } from './router/serviceWorker'

/**
 * Clears all API and SSR responses from the client cache
 * @return {Promise} resolved once the cache has been cleared
 */
export function clearClientCache() {
  return clearCache()
}
