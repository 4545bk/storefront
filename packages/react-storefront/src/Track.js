/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import { inject } from 'mobx-react'
import PropTypes from 'prop-types'
import analytics, { getTargets } from './analytics'

let nextId = 0
let ampAnalyticsTypes = {}

/**
 * @private
 * Used to reset the id for repeatable test results.
 */
export function resetId() {
  nextId = 0 
}

/**
 * Fires an analytics event when the user interacts with the child component.  By default this fires when the user
 * clicks on the child component, but this can be overriden using the `trigger` prop.  The value of trigger should
 * be the name of the event prop to bind to. All additional props will be passed as options along with the event.
 * 
 * Example:
 * 
 * <Track event="addedToCart" product={this.props.product}>
 *  <Button>Add to Cart</Button>
 * </Track>
 */
@inject('app')
export default class Track extends Component {
  
  constructor() {
    super()
    this.id = (nextId++).toString()
  }

  static propTypes = {
    /**
     * The name of the method to call on all configured analytics targets.
     * For example, "addedToCart".
     */
    event: PropTypes.string.isRequired,

    /**
     * The name of the handler prop on child component to intercept.  Defaults to "onClick"
     */
    trigger: PropTypes.string,

    /**
     * A function to call once the event has been successfully sent by all analytics targets.
     */
    onSuccess: PropTypes.func,

    /**
     * Additional data to send when tracking events in AMP.
     */
    ampData: PropTypes.object
  }

  static defaultProps = {
    trigger: 'onClick',
    onSuccess: Function.prototype,
    ampData: {}
  }

  componentWillMount() {
    if (this.props.app.amp) {
      const { event, trigger, app, children, onSuccess, ...data } = this.props
      this.createAmpTriggers(data)
    }
  }

  render() {
    return this.attachEvent()
  }

  /**
   * Fires the analytics event asynchronously using setImmediate so as not to 
   * block the render loop.
   */
  fireEvent() {
    const { event, trigger, app, children, onSuccess, ampData, ...data } = this.props

    setImmediate(async () => {
      await analytics.fire(event, data)
      onSuccess()
    })
  }

  /**
   * Intecepts the child's event handler prop corresponding to this.props.trigger
   * and returns the cloned child.
   * @return {React.Element}
   */
  attachEvent() {
    const { app: { amp }, trigger, children: el } = this.props

    if (el) {
      let originalHandler = el.props[trigger]

      const props = {
        ...el.props,
        [trigger]: (...args) => {
          if (originalHandler) originalHandler(...args)
          this.fireEvent()
        }
      }
  
      if (amp) {
        props['data-amp-id'] = this.id
        props['data-vars-moov-test'] = 'foo'
      }
  
      return React.cloneElement(el, props)
    } else {
      return null
    }
  }

  /**
   * Creates AMP event trigger objects based on this.props.event
   */
  createAmpTriggers(data) {
    const { event, children, ampData } = this.props

    for (let target of getTargets()) {
      const handler = target[event]
      const type = target.getAmpAnalyticsType()

      if (handler && type) {
        let eventData

        // Override send to capture the data that would be send instead of trying to send it
        target.send = data => eventData = data

        // Call the method corresponding to the event name, this should result in a call to send
        handler.call(target, data)

        if (eventData) {
          if (!eventData.trigger && children) {
            // no need to set a selector if we don't have a child element
            // an example of this is pageview events
            eventData.selector = `[data-amp-id="${this.id}"]`
          }
  
          this.configureAmpEvent(type, {
            on: eventData.selector ? 'click' : 'visible',
            ...eventData,
            request: 'event',
            ...ampData
          })
        } else {
          console.log(`WARNING: No data will be sent for the ${event} event when running in AMP because ${target.constructor.name} didn't return any data for this event.`)
        }
      }
    }
  }

  /**
   * Adds an AMP analytics event trigger
   * @param {String} type An AMP analytics event type, e.g. googleanalytics
   * @param {Object} trigger The trigger descriptor object
   */
  configureAmpEvent(type, trigger) {
    let config = ampAnalyticsTypes[type]

    if (!config) {
      config = ampAnalyticsTypes[type] = {
        triggers: []
      }
    }

    config.triggers.push(trigger)
  }

}

export function renderAmpAnalyticsTags() {
  const result = []

  for (let type in ampAnalyticsTypes) {
    result.push(
      `<amp-analytics${type === 'gtm' ? '' : ` type="${type}"`}>` +
        `<script type="application/json">${JSON.stringify(ampAnalyticsTypes[type])}</script>` +
      `</amp-analytics>`
    )
  }
  
  ampAnalyticsTypes = {}

  return result.join('\n')
}