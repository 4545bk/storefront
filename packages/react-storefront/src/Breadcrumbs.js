/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import ArrowRight from '@material-ui/icons/KeyboardArrowRight'
import Link from './Link'
import withStyles from '@material-ui/core/styles/withStyles'
import Typography from '@material-ui/core/Typography'
import { inject, observer } from 'mobx-react'
import Track from './Track'
import Container from './Container'

export const styles = theme => ({
  breadcrumbs: {
    margin: '0',
    backgroundColor: '#F4F2F1',
    padding: '12px 0',

    '& a': {
      color: theme.palette.text.primary,
      textDecoration: 'none'
    }
  },

  separator: {
    height: '12px',
    position: 'relative',
    top: '3px',
    width: '16px'
  },

  current: {
    fontWeight: 'bold',
    color: theme.palette.text.primary
  }
})

/**
 * Renders a list of breadcrumbs from AppModelBase.breadcrumbs.  When a breadcrumb is clicked, the breadcrumb_clicked
 * analytics event is automatically fired.
 */
@withStyles(styles, { name: 'RSFBreadcrumbs' })
@inject('app')
@observer
export default class Breadcrumbs extends Component {
  render() {
    const { app: { breadcrumbs }, classes } = this.props

    if (!breadcrumbs || !breadcrumbs.length) return null

    return (
      <Typography className={classes.breadcrumbs} variant="caption">
        <Container>
          { breadcrumbs.map(this.renderBreadcrumb) }
        </Container>
      </Typography>
    )
  }

  renderBreadcrumb = (item, i) => {
    const { classes } = this.props
    
    if (item.url) {
      return (
        <span key={i}>
          <Track event="breadcrumbClicked" breadcrumb={item}>
            <Link to={item.url}>{item.text}</Link>
          </Track>
          <ArrowRight className={classes.separator}/>
        </span>
      )
    } else {
      return (
        <span key={i} className={classes.current}>{item.text}</span>
      )
    }
  }

}
