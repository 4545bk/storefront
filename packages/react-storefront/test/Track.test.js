/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import { mount } from 'enzyme'
import { configureAnalytics } from '../src/analytics'
import waitForAnalytics from './helpers/waitForAnalytics'
import Track, { renderAmpAnalyticsTags, resetId } from '../src/Track'
import TestProvider from './TestProvider'

describe('Track', () => {

  beforeEach(() => {
    resetId()
  })

  it('should fire the configured event when clicked', () => {
    const testEvent = jest.fn()

    configureAnalytics({ testEvent })

    mount(
      <TestProvider>
        <Track event="testEvent" foo="bar">
          <button>Click Me</button>
        </Track>
      </TestProvider>
    )
      .find('button')
      .simulate('click')

    return waitForAnalytics(() => {
      expect(testEvent).toHaveBeenCalledWith({ foo: 'bar' })
    })
  })

  it('should call the onSuccess prop after the event has been sent', (done) => {
    const testEvent = jest.fn()
    const onSuccess = jest.fn()

    configureAnalytics({ testEvent })

    mount(
      <TestProvider>
        <Track event="testEvent" foo="bar" onSuccess={onSuccess}>
          <button>Click Me</button>
        </Track>
      </TestProvider>
    )
      .find('button')
      .simulate('click')

    setTimeout(() => {
      expect(onSuccess).toHaveBeenCalled()
      done()
    }, 200)
  })
  
  it('calls the original handler prop', () => {
    const testEvent = jest.fn(), onClick = jest.fn()

    configureAnalytics({ testEvent })

    mount(
      <TestProvider>
        <Track event="testEvent" foo="bar">
          <button onClick={onClick}>Click Me</button>
        </Track>
      </TestProvider>
    )
      .find('button')
      .simulate('click')

    expect(onClick).toHaveBeenCalledWith(expect.any(Object))
  })

  it('should create AMP triggers', () => {
    configureAnalytics({
      testEvent() {},
      getAmpAnalyticsType() {
        return "test"
      },
      getPropsForAmpAnalytics() {
        return {
          request: 'event',
          vars: {
            eventCategory: 'search',
            eventAction: 'click'
          }
        }
      }
    })

    const wrapper = mount(
      <TestProvider app={{ amp: true }}>
        <Track event="testEvent" foo="bar">
          <button>Click Me</button>
        </Track>
      </TestProvider>
    )

    // check the amp trigger
    expect(renderAmpAnalyticsTags()).toEqual(
      '<amp-analytics type="test">' +
        '<script type="application/json">' +
          '{"triggers":[{"on":"click","selector":"[data-amp-id=\\"0\\"]","request":"event","vars":{"eventCategory":"search","eventAction":"click"}}]}' +
        '</script>' +
      '</amp-analytics>'
    )

    // make sure it adds the data-amp-id to the wrapped element
    expect(wrapper.find('button[data-amp-id="0"]').length).toBe(1)
  })
})