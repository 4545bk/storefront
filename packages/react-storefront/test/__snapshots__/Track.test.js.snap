// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track should support no children 1`] = `
<TestProvider>
  <Provider router={{...}} app={{...}} history={{...}}>
    <inject-Track-with-app event="testEvent" foo="bar">
      <Track event="testEvent" foo="bar" app={{...}} trigger="onClick" onSuccess={[Function]} />
    </inject-Track-with-app>
  </Provider>
</TestProvider>
`;
